File: src/bumperbot_utils/bumperbot_utils/trajectory_drawer.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from nav_msgs.msg import Odometry, Path
from geometry_msgs.msg import PoseStamped

class TrajectoryDrawer(Node):
    def __init__(self):
        super().__init__("trajectory_drawer")

        self.odom_sub_ = self.create_subscription(Odometry, "bumperbot_controller/odom", self.odomCallback, 10)
        self.traj_pub_ = self.create_publisher(Path, "bumperbot_controller/trajectory", 10)

        self.traj_msg_ = Path()
        self.traj_msg_.header.frame_id = "odom"

    def odomCallback(self, odom_msg: Odometry):
        pose = odom_msg.pose.pose

        pose_stamped = PoseStamped()
        pose_stamped.pose = pose
        pose_stamped.header.stamp = self.get_clock().now().to_msg()
        pose_stamped.header.frame_id = "odom"

        self.traj_msg_.header.stamp = self.get_clock().now().to_msg()
        self.traj_msg_.poses.append(pose_stamped)

        self.traj_pub_.publish(self.traj_msg_)

def main():
    rclpy.init()

    trajectory_drawer = TrajectoryDrawer()
    rclpy.spin(trajectory_drawer)
    trajectory_drawer.destroy_node()

    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/bumperbot_utils/src/trajectory_drawer.cpp
#include "bumperbot_utils/trajectory_drawer.hpp"

using std::placeholders::_1;

TrajectoryDrawer::TrajectoryDrawer() : Node("trajectory_drawer")
{
    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>("bumperbot_controller/odom", 10, std::bind(&TrajectoryDrawer::odomCallback, this, _1));
    traj_pub_ = create_publisher<nav_msgs::msg::Path>("bumperbot_controller/trajectory", 10);

    traj_msg_.header.frame_id = "odom";
}

void TrajectoryDrawer::odomCallback(const nav_msgs::msg::Odometry &odom_msg)
{
    auto pose_stamped = geometry_msgs::msg::PoseStamped();
    pose_stamped.pose = odom_msg.pose.pose;
    pose_stamped.header.stamp = get_clock()->now();
    pose_stamped.header.frame_id = "odom";

    traj_msg_.header.stamp = get_clock()->now();
    traj_msg_.poses.push_back(pose_stamped);

    traj_pub_->publish(traj_msg_);
}

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<TrajectoryDrawer>();
    rclcpp::spin(node);

    rclcpp::shutdown();

    return 0;
}